LightGBM Unbalanced Data Argument [Issue #3688 Fix]

RadicalRayan opened the pull request:
Fix for issue #3688. LightGBM Multiclass Trainer can now accept unbalanced data parameter as was previously possible in the Binary Trainer. An additional argument was added to the LightGBMBinaryEstimator test.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=3925) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=3925) before we can accept your contribution.<br/><br/>:x: RadicalRayan [sign now](https://cla.dotnetfoundation.org/dotnet/machinelearning?pullRequest=3925)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/machinelearning?pullRequest=3925) it.</sub>

wschin commented on src/Microsoft.ML.LightGbm/LightGbmArguments.cs:
```diff
@@ -58,6 +59,11 @@ public BoosterParameterBase(OptionsBase options)
+            /// Whether training data is unbalanced. Used by <see cref="LightGbmBinaryTrainer"/>.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Use for binary classification when training data is not balanced.", ShortName = "us")]
+            public bool UnbalancedSets = false;
```
We need two tests for `UnbalancedSets=true` and `UnbalancedSets=false`. In addition to simply run a model, we need more restrictive comparison like
```c#
        [LightGBMFact]
        public void LightGbmMulticlassEstimatorCompareOva()
        {
            float sigmoidScale = 0.5f; // Constant used train LightGBM. See gbmParams["sigmoid"] in the helper function.

            // Train ML.NET LightGBM and native LightGBM and apply the trained models to the training set.
            LightGbmHelper(useSoftmax: false, sigmoid: sigmoidScale, out string modelString, out List<GbmExample> mlnetPredictions, out double[] nativeResult1, out double[] nativeResult0);

            // The i-th predictor returned by LightGBM produces the raw score, denoted by z_i, of the i-th class.
            // Assume that we have n classes in total. The i-th class probability can be computed via
            // p_i = sigmoid(sigmoidScale * z_i) / (sigmoid(sigmoidScale * z_1) + ... + sigmoid(sigmoidScale * z_n)).
            Assert.True(modelString != null);
            // Compare native LightGBM's and ML.NET's LightGBM results example by example
            for (int i = 0; i < _rowNumber; ++i)
            {
                double sum = 0;
                for (int j = 0; j < _classNumber; ++j)
                {
                    Assert.Equal(nativeResult0[j + i * _classNumber], mlnetPredictions[i].Score[j], 6);
                    if (float.IsNaN((float)nativeResult1[j + i * _classNumber]))
                        continue;
                    sum += MathUtils.SigmoidSlow(sigmoidScale * (float)nativeResult1[j + i * _classNumber]);
                }
                for (int j = 0; j < _classNumber; ++j)
                {
                    double prob = MathUtils.SigmoidSlow(sigmoidScale * (float)nativeResult1[j + i * _classNumber]);
                    Assert.Equal(prob / sum, mlnetPredictions[i].Score[j], 6);
                }
            }

            Done();
        }
```
in `machinelearning2\test\Microsoft.ML.Tests\TrainerEstimators\TreeEstimators.cs`. Note that this test directly checks if ML.NET's LightGBM module and LightGBM's C API produce identical numbers.

RadicalRayan replied:
Thank you. I will add and run these tests

RadicalRayan commented:
Needed to add argument to LightGbmMulticlassTrainer. Added new test for unbalanced binary tree, added two tests for unbalanced/balanced multiclass, added a test to compare multiclass unbalanced .NET with C. Revision to manifest file.

eerhardt commented on src/Microsoft.ML.LightGbm/LightGbmMulticlassTrainer.cs:
```diff
@@ -75,6 +75,12 @@ public enum EvaluateMetricType
+            /// <summary>
+            /// Whether training data is unbalanced.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Use for binary classification when training data is not balanced.", ShortName = "us")]
```
> Use for binary classification when training data is not balanced.

It's odd to say "use for binary classification" when this is the multiclass trainer.

RadicalRayan replied:
Is it more helpful to just say "Use for multi-class classification when training data is not balanced."

eerhardt commented on src/Microsoft.ML.LightGbm/LightGbmArguments.cs:
```diff
@@ -58,6 +59,11 @@ public BoosterParameterBase(OptionsBase options)
+            /// <summary>
+            /// Whether training data is unbalanced. Used by <see cref="LightGbmBinaryTrainer"/>.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Use for binary classification when training data is not balanced.", ShortName = "us")]
```
> Use for binary classification when training data is not balanced.

Is putting this option on the `OptionsBase` class correct if it is only for binary classification?

RadicalRayan replied:
You're right. The cause of the problem was this [commit](https://github.com/dotnet/machinelearning/pull/2948/files#diff-8173688cd56a37cbcc92012f7d22a31dL152) which split the trainers in to their classes (binary and multiclass). is it better not to include it as an argument in this file?

eerhardt commented on src/Microsoft.ML.LightGbm/LightGbmArguments.cs:
```diff
@@ -58,6 +59,11 @@ public BoosterParameterBase(OptionsBase options)
+            /// Whether training data is unbalanced. Used by <see cref="LightGbmBinaryTrainer"/>.
+            /// </summary>
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Use for binary classification when training data is not balanced.", ShortName = "us")]
+            public bool UnbalancedSets = false;
```
Where is this boolean being used? I don't see any code added using it.

RadicalRayan replied:
It's added to the dictionary, NameMapping on line 39:
{nameof(OptionsBase.UnbalancedSets),                 "is_unbalance"},

eerhardt replied:
You are declaring 2 new boolean fields. I only see 1 being used.

codemzs commented on src/Microsoft.ML.LightGbm/LightGbmMulticlassTrainer.cs:
```diff
@@ -110,6 +116,7 @@ static Options()
             {
                 var res = base.ToDictionary(host);
 
+                res[GetOptionName(nameof(UnbalancedSets))] = UnbalancedSets;
```
>res[GetOptionName(nameof(UnbalancedSets))] = UnbalancedSets; [](start = 16, length = 60)

how is getting mapped to is_unbalance 🔗︎, default = false, type = bool, aliases: unbalance, unbalanced_sets

codemzs replied:
nevermind I see it now:

public static string GetOptionName(string name)
        {
            // Otherwise convert the name to the light gbm argument
            StringBuilder strBuf = new StringBuilder();
            bool first = true;
            foreach (char c in name)
            {
                if (char.IsUpper(c))
                {
                    if (first)
                        first = false;
                    else
                        strBuf.Append('_');
                    strBuf.Append(char.ToLower(c));
                }
                else
                    strBuf.Append(c);
            }
            return strBuf.ToString();
        }

---
In reply to: [299023864](https://github.com/dotnet/machinelearning/pull/3925#discussion_r299023864) [](ancestors = 299023864)

wschin commented on test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs:
```diff
@@ -372,7 +432,9 @@ private void LightGbmHelper(bool useSoftmax, double sigmoid, out string modelStr
-                    Sigmoid = sigmoid // Custom sigmoid value.
+                    Sigmoid = sigmoid, // Custom sigmoid value.
+                    UnbalancedSets = unbalancedSets // false by default
+                    
```
Empty line.
